File: code/GestureworksCore/bindings/c++/OpenFrameworks/examples/CollectionViewer/collectionViewer.cpp
////////////////////////////////////////////////////////////////////////////////
//
//  IDEUM
//  Copyright 2011-2013 Ideum
//  All Rights Reserved.
//
//  GestureWorks
//
//  File:    collectionViewer.cpp
//  Authors:  Ideum
//             
//  NOTICE: Ideum permits you to use, modify, and distribute this file only
//  in accordance with the terms of the license agreement accompanying it.
//
////////////////////////////////////////////////////////////////////////////////

#pragma once

#include "ofMain.h"
#include "collectionViewer.h"
#include "MediaBoxes.h"

//--------------------------------------------------------------
void collectionViewer::setup(){
	ofBackground(255,255,255);
	window_position_x = ofGetWindowPositionX();
	window_position_y = ofGetWindowPositionY();

	background.loadImage("gw_c.png");

	ofEnableAlphaBlending();
	ofSetFullscreen(true);

	last_tick_count = GetTickCount();


		//Create objects that we'll be using
		PictureBox * boccioni = new PictureBox(this, scene_list, "boccioni", 331, 600);
		boccioni->loadImage("boccioni_600.png");
		boccioni->loadToggle("boccioni_600_reverse.png");
		boccioni->setRotation(3.1415926/4);
		boccioni->setScale(1);

		MovieBox * clock = new MovieBox(this, scene_list, "clock", 720, 404);
		clock->loadMovie("clock.mp4");
		clock->loadToggle("clock_back.png");
		clock->setX(700);
		clock->setY(300);
		clock->setScale(1);

		scene_list.push_back(boccioni);
		scene_list.push_back(clock);

		vector<MediaBox*>::iterator scene_it;


	if(loadGestureWorks("GestureWorksCore.dll")) { std::cout << "Error loading gestureworks dll" << std::endl; exit(); }
	if(!loadGML("collection_viewer_gestures.gml")) { std::cout << "Could not find gml file" << std::endl; exit(); }
	initializeGestureWorks(1920,1080);

	if( !registerWindowForTouchByName("testify") ) { std::cout << "Could not register collection viewer window for touch." << std::endl; exit(); }

	registerTouchObject("boccioni");
	registerTouchObject("boccioni.toggle");
	registerTouchObject("boccioni.close");
	registerTouchObject("clock");
	registerTouchObject("clock.play");
	registerTouchObject("clock.stop");
	registerTouchObject("clock.toggle");
	registerTouchObject("clock.close");

	//Set up and register boccioni picturebox
	scene_map["boccioni"] = (MediaBox*)boccioni;
	scene_map["boccioni.toggle"] = (MediaBox*)&boccioni->info_toggle;
	scene_map["boccioni.close"] = (MediaBox*)&boccioni->close;

	addGesture("boccioni", "n-drag");
	addGesture("boccioni", "n-rotate");
	addGesture("boccioni", "n-scale");

	addGesture("boccioni.toggle", "n-tap");
	addGesture("boccioni.close", "n-tap");

	//Set up and register clock moviebox
	scene_map["clock"] = (MediaBox*)clock;
	scene_map["clock.play"] = (MediaBox*)&clock->play;
	scene_map["clock.stop"] = (MediaBox*)&clock->stop;
	scene_map["clock.toggle"] = (MediaBox*)&clock->info_toggle;
	scene_map["clock.close"] = (MediaBox*)&clock->close;

	addGesture("clock", "n-drag");
	addGesture("clock", "n-rotate");
	addGesture("clock", "n-scale");

	addGesture("clock.play", "n-tap");
	addGesture("clock.stop", "n-tap");
	addGesture("clock.toggle", "n-tap");
	addGesture("clock.close", "n-tap");


}





//--------------------------------------------------------------
void collectionViewer::keyPressed  (int key){
    
}

//--------------------------------------------------------------
void collectionViewer::keyReleased(int key){

}

//--------------------------------------------------------------
void collectionViewer::mouseMoved(int x, int y ){

}

//--------------------------------------------------------------
void collectionViewer::mouseDragged(int x, int y, int button){

}

//--------------------------------------------------------------
void collectionViewer::mousePressed(int x, int y, int button){

}


//--------------------------------------------------------------
void collectionViewer::mouseReleased(int x, int y, int button){

}

//--------------------------------------------------------------
void collectionViewer::windowResized(int w, int h){

}

//--------------------------------------------------------------
void collectionViewer::gotMessage(ofMessage msg){

}

//--------------------------------------------------------------
void collectionViewer::dragEvent(ofDragInfo dragInfo){ 

}
	
//--------------------------------------------------------------
void collectionViewer::draw(){

	background.draw(0, 0, 1920, 1080);

	vector<MediaBox*>::iterator scene_it;
	for(scene_it = scene_list.begin(); scene_it < scene_list.end(); scene_it++) {
		(*scene_it)->draw();
	}
}


void collectionViewer::update(){

	if(GetTickCount() - last_tick_count < 16) {
		return;
	} else { last_tick_count = GetTickCount(); }
    vector<MediaBox*>::iterator scene_it;

	processFrame();

	std::vector<gwc::PointEvent> point_events = consumePointEvents();

	
	for(std::vector<gwc::PointEvent>::iterator event_it = point_events.begin(); event_it != point_events.end(); event_it++) {
		if(event_it->status == gwc::TOUCHADDED) {
			std::string hit_object = hitTest(event_it->position.getX(), event_it->position.getY());
			if(hit_object != "") {
				assignTouchPoint(hit_object, event_it->point_id);
				std::cout << "Assigned touchpoint to " << hit_object << std::endl;
			}
		}
	}

	std::vector<gwc::GestureEvent> gesture_events = consumeGestureEvents();
	for(std::vector<gwc::GestureEvent>::iterator event_it = gesture_events.begin(); event_it != gesture_events.end(); event_it++) {
		std::map<std::string,MediaBox*>::iterator temp = scene_map.find(event_it->target);
		if(temp != scene_map.end()) {
			temp->second->handleEvent(*event_it);
		}
	}


	for(scene_it = scene_list.begin(); scene_it < scene_list.end(); scene_it++) {
		(*scene_it)->update();
	}
}

//Perform a hit test for a received point
std::string collectionViewer::hitTest(float x, float y) {
		
		vector<MediaBox*>::reverse_iterator scene_it;
		for(scene_it = scene_list.rbegin(); scene_it < scene_list.rend(); scene_it++)
		{
			if(!(*scene_it)->children.empty()) {
				vector<MediaBox*>::iterator child_it;
				for(child_it = (*scene_it)->children.begin(); child_it < (*scene_it)->children.end(); child_it++) {
					if(test_point(x, y, (*child_it)->x, (*child_it)->y, (*child_it)->width, (*child_it)->height, (*child_it)->rotation, (*child_it)->scale)) {
						return (*child_it)->object_id;
					}
				}
			}
			if(test_point(x, y, (*scene_it)->x, (*scene_it)->y, (*scene_it)->width, (*scene_it)->height, (*scene_it)->rotation, (*scene_it)->scale)) {
				return (*scene_it)->object_id;
			}
		}
		return "";
}


bool test_point(float point_x, float point_y, float box_x, float box_y, float box_width, float box_height, float box_angle, float box_scale) {
	
	float local_x = rotateAboutCenterX(point_x, point_y, box_x, box_y, -box_angle);
	float local_y = rotateAboutCenterY(point_x, point_y, box_x, box_y, -box_angle);
	
	if( (abs(local_x - box_x) <= box_width*box_scale/2) && (abs(local_y - box_y) <= box_height*box_scale/2) ) {
		return true;
	}
	return false;
}

float radsToDegrees(float rad) {
	return rad * 180 / 3.14159265358979323846264338;
}

float degreesToRads(float deg) {
	return 3.14159265358979323846264338 * deg / 180;
}

//Utility functions for doing coordinate transforms

float rotateAboutCenterX(float point_x, float point_y, float center_x, float center_y, float ref_angle) {

	float local_x = point_x - center_x;
	float local_y = point_y - center_y;
		
	float length = sqrt(local_x*local_x + local_y*local_y);
		
	return (length * cos(ref_angle + atan2(local_y, local_x)) + center_x);
}
	
float rotateAboutCenterY(float point_x, float point_y, float center_x, float center_y, float ref_angle) {

	float local_x = point_x - center_x;
	float local_y = point_y - center_y;
		
	float length = sqrt(local_x*local_x + local_y*local_y);
		
	return point_y = length * sin(ref_angle + atan2(local_y, local_x)) + center_y;
}
	
	
float getDrawingCoordX(float width, float height, float box_x, float box_y, float box_angle, float box_scale) {
	float length = sqrt(box_scale*box_scale*(width*width + height*height)/4);
	return (length * cos(box_angle + atan2(-height, -width)) + box_x);
}
	
float getDrawingCoordY(float width, float height, float box_x, float box_y, float box_angle, float box_scale) {
	float length = sqrt(box_scale*box_scale*(width*width + height*height)/4);
	return (length * sin(box_angle + atan2(-height, -width)) + box_y);
}

File: code/GestureworksCore/bindings/python/Kivy/examples/CollectionViewer/collectionviewer.py
'''
////////////////////////////////////////////////////////////////////////////////
//
//  IDEUM
//  Copyright 2011-2013 Ideum
//  All Rights Reserved.
//
//  Gestureworks Core
//
//  File:    collectionviewer.py
//  Authors:  Ideum
//
//  NOTICE: Ideum permits you to use, modify, and distribute this file only
//  in accordance with the terms of the license agreement accompanying it.
//
////////////////////////////////////////////////////////////////////////////////
'''


from gwc_python.core import GestureWorksCore
from gwc_python.GWCUtils import TOUCHADDED, TOUCHREMOVED, TOUCHUPDATE, rotateAboutCenter

import kivy
kivy.require('1.2.0')
 
from kivy.app import App
from kivy.uix.widget import NumericProperty, StringProperty
from kivy.uix.image import Image
from kivy.uix.scatter import Scatter
from kivy.uix.video import Video
from kivy.factory import Factory
from kivy.graphics.transformation import Matrix
from kivy.clock import Clock

from math import radians
from os.path import join
from os import getcwd
from random import randint


"""This will be the container for all of our touch objects"""
class TouchObject(Scatter):
    source = StringProperty(None)
    frame_pad = NumericProperty(None)
    
    # override kivy touch handling
    def on_touch_move(self, touch): return
    
class CollectionApp(App):
    
    """Initialize data.
    
    Takes a GestureWorksCore object
    
    """
    def __init__(self, gw):
        self.gw = gw
        self.objects = {}
        self.active_touches = {}
        self.last_active = []
        self.touch_points = {}

        Clock.schedule_interval(self.updateGW, 1./60.)
        super(CollectionApp, self).__init__()
        
    
    """Redraw a widget as the topmost element of its parent"""
    def _draw_on_top(self, widg, parent):
        parent.remove_widget(widg)
        parent.add_widget(widg)
    
    
    """Check if a point given by x and y has collided with an object
    
    Returns True if there was a collision
    
    """
    def hitTest(self, obj, x, y):
        
        y = self.root.height - y
        (local_x, local_y) = rotateAboutCenter(x, y, obj.center_x, obj.center_y, radians(-obj.rotation))

        if obj.name == 'clock':
            local_y += 60 * obj.scale # offset because buttons hang off the video widget

        return (abs(local_x - obj.center_x) <= obj.width*obj.scale/2) and (abs(local_y - obj.center_y) <= obj.height*obj.scale/2)   
    
    
    """Process each touch event from GestureWorks and check if new points should be added
    to touch objects.
    
    """    
    def processTouchEvents(self, touches):
        for touch in touches:
            if touch.status == TOUCHADDED:
                self.touch_points.update({touch.point_id: touch})
                hits = []
                for name in self.objects.keys():
                    obj = self.objects[name]['obj']
                    if not isinstance(obj, Scatter): continue 
                    if self.hitTest(obj, touch.position.x, touch.position.y):
                        hits.append(name)
                        
                # If both objects were hit, decide which one is on top
                if len(hits) > 1:
                    for hit in reversed(hits):
                        if self.objects[hit]['obj'].on_top:
                            self.gw.addTouchPoint(hit, touch.point_id)
                            
                elif len(hits) == 1:
                    for o in self.objects.values(): o['obj'].on_top = False
                    self.objects[hits[0]]['obj'].on_top = True
                    self._draw_on_top(self.objects[hits[0]]['obj'], self.root)
                    self.gw.addTouchPoint(hits[0], touch.point_id)       
                                
            elif touch.status == TOUCHUPDATE:
                self.touch_points.update({touch.point_id: touch})
            elif touch.status == TOUCHREMOVED:
                self.touch_points.pop(touch.point_id)
             
            
    """Translate the object based on the event deltas"""
    def handleDrag(self, obj, e): 
        obj.center_x += e.values['drag_dx']
        obj.center_y -= e.values['drag_dy']
        
        # Check window boundaries
        obj.center_x = max(min(obj.center_x, self.root.width), 0)
        obj.center_y = max(min(obj.center_y, self.root.height), 0)     
    
    
    """Rotate the object around the center of the event"""
    def handleRotate(self, obj, gesture_event):
        theta = gesture_event.values['rotate_dtheta']
        obj.rotation -= theta
        if gesture_event.n:
            obj.center = rotateAboutCenter(obj.center_x, obj.center_y, gesture_event.x, self.root.height - gesture_event.y, radians(-theta))
    
    """Scale the object"""
    def handleScale(self, obj, e):
        dsx = e.values['scale_dsx']
        obj.scale += dsx
        obj.scale = max(min(obj.scale, 2), .5) # Scale bounds
                
    """Check for and process button taps"""
    def handleTap(self, obj, e):
        x = e.values['tap_x']
        y = self.root.height - e.values['tap_y']
        local_x, local_y = obj.to_widget(x, y)
        for child in obj.children:
            try:
                child.action # only care about objects with action functions
            except AttributeError:
                continue
            
            w = child.width * obj.scale
            h = child.height * obj.scale
            if abs(local_x - child.center_x) <= w/2 and abs(local_y - child.center_y) <= h/2:
                child.action(e)
                return
            
                
    """Get point and gesture event data from GestureWorks"""
    def updateGW(self, *args):
        self.gw.processFrame() # tell GW to process a frame of data
        point_events = gw.consumePointEvents() # get the point events from that frame
        gesture_events = gw.consumeGestureEvents() # get the gesture events from that frame
        self.processTouchEvents(point_events)

        # Call event handlers
        for event in gesture_events:
            obj = self.objects[event.target]['obj']
            {'n-drag': self.handleDrag,
             'n-rotate': self.handleRotate,
             'n-scale': self.handleScale,
             'n-tap': self.handleTap}[event.gesture_id](obj, event)

    
    """Remove a widget when its "close" button is tapped."""
    def handleCloseWidget(self, evt):
        obj = self.objects[evt.target]['obj']
        self.root.remove_widget(obj)
     
    
    """Switch a container's image with its alternate and vice versa."""
    def handleMetaSwitch(self, evt):
        obj = self.objects[evt.target]['obj']

        container = obj
        alt = container.alt
        pic = container.main
        
        # shuffle images to appear to "switch"
        self._draw_on_top(alt, container)
        self._draw_on_top(container.alt_info, container)
        container.remove_widget(container.info)
        
        info = container.info
        container.info = container.alt_info
        container.alt_info = info
        
        container.alt = pic
        container.remove_widget(pic)
        container.main = alt

    
    """Toggle a video between play and pause modes."""
    def handlePause(self, evt):  
        obj = self.objects[evt.target]['obj']
        
        if not isinstance(obj.main, Video): return # Disable control if meta image is displayed
        obj.main.play = not obj.main.play
        playing = obj.main.play
        parent = obj
        play = parent.play
        pause = parent.pause
        if playing:
            parent.remove_widget(play)
            parent.add_widget(pause)
        else:
            parent.remove_widget(pause)
            parent.add_widget(play)

       
    """Stop a video by setting its current position to the beginning and pausing it.""" 
    def handleStop(self, evt):
        obj = self.objects[evt.target]['obj']
        video = obj.main
        if not isinstance(video, Video): return # Disable control if meta image is displayed
        video.play = False
        video.seek(0)
        obj.remove_widget(obj.pause)
        obj.add_widget(obj.play)
    
    
    
    """Add a button to a parent widget"""
    def _addButton(self, parent, name, source, size, pos, action_func, on_top=False):
        button = Image(source=source)
        button.size = size
        button.on_top = on_top
        button.pos = pos
        button.scale = 1.
        button.rotation = 0.
        button.action = action_func
        button.name = name
        self.objects.update({name: {'obj': button}})
        parent.add_widget(button)
        
        return button
    
    
    """Create a container widget extending the Scatter class that will respond to touch events and display
    an image or video. Will automatically add container and buttons to internal structures and register
    callbacks for event listeners.
    
    """
    def _create_collection_item(self, source, alt_source, size, position, frame_size, name, is_video=False):
        container = TouchObject(frame_pad=frame_size)
        
        container.main = {True: Video, False: Image}[is_video]
        container.on_top = is_video
        container.main = container.main(source=source)
        container.pos = position
    
        container.main.size = size[0], size[1] / container.main.image_ratio
        container.size = container.main.size
        
        container.alt = Image(source=alt_source)
        container.alt.size = container.main.size

        container.add_widget(container.main)

        button_y = [0, -60][is_video]
        
        info_pos = (container.main.width-100, button_y)
    
        container.name = name
        container.alt_info = self._addButton(container, name + '_alt_info_button', 'media/images/GW_return.png', (40,70), info_pos, self.handleMetaSwitch)
        container.remove_widget(container.alt_info)
        container.info = self._addButton(container, name + '_info_button', 'media/images/GW_info.png', (40,70), info_pos, self.handleMetaSwitch, True)
        container.close = self._addButton(container, name + '_close_button', 'media/images/GW_close.png', (40,70), (container.main.width-55, button_y), self.handleCloseWidget, True)
        self.objects.update({name: {'obj': container}})
           
        if is_video:
            container.pause = self._addButton(container, name + '_pause_button', 'media/images/GW_pause.png', (70,70), (-10, button_y), self.handlePause)
            container.play = self._addButton(container, name + '_play_button', 'media/images/GW_play.png', (70,70), (-10, button_y), self.handlePause)
            container.remove_widget(container.play)
            container.stop = self._addButton(container, name + '_stop_button', 'media/images/GW_stop.png', (70,70), (35, button_y), self.handleStop)       
        
        return container
        

    """Setup and register the touch objects"""
    def build(self):
        if not self.gw.registerWindow('Kivy'): exit()
        
        curdir = getcwd()
        image_file = join(curdir, 'media/images/pics/boccioni_600.png')
        alt_image_file = join(curdir, 'media/images/pics/meta/boccioni_600.png')
        image_name = 'image'
        img = self._create_collection_item(image_file, alt_image_file, (400, 400), (400, 400), 0, image_name, False)
        img.rotation = randint(-45, 45)
        self.root.add_widget(img)
        self.gw.registerTouchObject(image_name)
        self.gw.addGesture(image_name, 'n-drag')
        self.gw.addGesture(image_name, 'n-rotate')
        self.gw.addGesture(image_name, 'n-scale')
        self.gw.addGesture(image_name, 'n-tap')
        
        video_file = join(curdir, 'media/videos/clock.mp4')
        alt_video_file = join(curdir, 'media/videos/meta/clock.png')
        video_name = 'clock'
        vid = self._create_collection_item(video_file, alt_video_file, (400,225), (400,400), 0, video_name, True)
        self.root.add_widget(vid)
        vid.main.play = True
        self.gw.registerTouchObject(video_name)
        self.gw.addGesture(video_name, 'n-drag')
        self.gw.addGesture(video_name, 'n-rotate')
        self.gw.addGesture(video_name, 'n-scale')
        self.gw.addGesture(video_name, 'n-tap')
        
Factory.register('TouchObject', TouchObject)
    
if __name__ == '__main__':
    bin_path = "C:\\gwdistro\\dev\\GestureworksCore\\"
    gw = GestureWorksCore(dll_path=bin_path + 'GestureworksCore32.dll')
    gw.initializeGestureWorks()
    try: 
        gw.loadGML(bin_path + 'basic_manipulation.gml')
    except WindowsError, e:
        print 'Unable to load GML from', bin_path 
    
    collection = CollectionApp(gw)
    
    collection.run()
    